### ISSUE TITLE
Reports packages as behind, even if they are not.

### ISSUE THREAD

**Issue by andersaloof** [tjunnone/npm-check-updates#185]
Just updated to 2.5.4 and ran 'ncu'.

Got a message that a bunch of packages was behind, so I ran npm update. 

 autoprefixer                    ^6.0.3  →   ^6.1.2 
 browser-sync                   ^2.9.10  →  ^2.10.0 
 browserify-shim                ^3.8.10  →  ^3.8.11 
 counterpart                    ^0.17.0  →  ^0.17.1 
 csswring                        ^4.0.0  →   ^4.1.1 
 del                             ^2.0.2  →   ^2.2.0 
 es6-promise-polyfill            ^1.0.3  →   ^1.1.1 
 gulp-debug                      ^2.1.1  →   ^2.1.2 
 gulp-git                        ^1.5.0  →   ^1.6.0 
 gulp-sass                       ^2.0.4  →   ^2.1.0 
 gulp-uglify                     ^1.4.1  →   ^1.5.1 
 gulp-util                       ^3.0.6  →   ^3.0.7 
 history                        ^1.13.1  →  ^1.14.0 
 react                          ^0.14.2  →  ^0.14.3 
 react-addons-test-utils        ^0.14.2  →  ^0.14.3 
 react-addons-transition-group  ^0.14.2  →  ^0.14.3 
 react-addons-update            ^0.14.2  →  ^0.14.3 
 react-dom                      ^0.14.2  →  ^0.14.3 
 react-google-maps               ^4.3.1  →   ^4.6.1 
 react-router                    ^1.0.0  →   ^1.0.1 
 run-sequence                    ^1.1.4  →   ^1.1.5 
 sinon                          ^1.17.1  →  ^1.17.2 
 superagent                      ^1.4.0  →   ^1.5.0 
 watchify                        ^3.4.0  →   ^3.6.1 

Ran 'npm update', but after a successfull update running 'ncu' still reports that the same packages are behind, even though 'npm ls --depth=0' reports that the packages has the correct updated versions.

├── autoprefixer@6.1.2
├── babel-core@5.8.34
├── babel-jest@5.3.0
├── babel-plugin-react-transform@1.1.1
├── babelify@6.4.0
├── bows@1.4.8
├── browser-sync@2.10.0
├── browserify@11.2.0
├── browserify-shim@3.8.11
├── counterpart@0.17.1
├── csswring@4.1.1
├── del@2.2.0
├── es6-promise-polyfill@1.1.1
├── es6-shim@0.33.13
├── geolib@2.0.18
├── google-map-react@0.9.3
├── gulp@3.9.0
├── gulp-babel@5.3.0
├── gulp-changed@1.3.0
├── gulp-composer@0.1.0
├── gulp-concat@2.6.0
├── gulp-debug@2.1.2
├── gulp-expect-file@0.0.7
├── gulp-filesize@0.0.6
├── gulp-filter@3.0.1
├── gulp-git@1.6.0
├── gulp-modernizr@0.0.0
├── gulp-notify@2.2.0
├── gulp-phpunit@0.9.3
├── gulp-plumber@1.0.1
├── gulp-postcss@6.0.1
├── gulp-rename@1.2.2
├── gulp-replace@0.5.4
├── gulp-sass@2.1.0
├── gulp-sequence@0.4.1
├── gulp-sourcemaps@1.6.0
├── gulp-task-listing@1.0.1
├── gulp-uglify@1.5.1
├── gulp-util@3.0.7
├── history@1.14.0
├── jest-cli@0.5.10
├── livereactload@2.1.0
├── localforage@1.3.0
├── lockdown@0.0.8-dev
├── lodash@3.10.1
├── lodash.clone@3.0.3
├── lodash.isequal@3.0.4
├── lodash.merge@3.3.2
├── lodash.throttle@3.0.4
├── object-assign@4.0.1
├── pretty-hrtime@1.0.1
├── pubsub-js@1.5.3
├── raf@3.1.0
├── react@0.14.3
├── react-addons-test-utils@0.14.3
├── react-addons-transition-group@0.14.3
├── react-addons-update@0.14.3
├── react-dom@0.14.3
├── react-google-maps@4.6.1
├── react-h5-video@1.0.0
├── react-icons@0.1.5
├── react-proxy@1.1.1
├── react-responsive@0.0.10
├── react-router@1.0.1
├── reflux@0.3.0
├── require-dir@0.3.0
├── run-sequence@1.1.5
├── shortid@2.2.4
├── sinon@1.17.2
├── superagent@1.5.0
├── superagent-bluebird-promise@2.1.1
├── vinyl-buffer@1.0.0
├── vinyl-source-stream@1.1.0
└── watchify@3.6.1

I'm on node 5.1.0 and Windows...

**Comment by metaraine** [comment]
Hi Anders, thanks for reporting. Are the versions [not updating because they are satisfied by the specified range](https://github.com/tjunnone/npm-check-updates#why-is-it-not-updating-100-to-101-when-101-is-the-latest)?

**Comment by jlbruno** [comment]
You ran `npm update` which updated your packages, but did you run `ncu -u` to update the versions listed in package.json?

**Comment by andersaloof** [comment]
I may be wrong here, but it seems to me that something changed in the latest update, that doesn't correctly detect the latest installed version of a package. To clarify, this issue only applies to packages which is within the declared version range (as I don't update the package.json file at any point).

I reverted the installed version to 2.5.1 now and get this output.

```
 babel-core        ^5.8.25  →      ^6.3.15 
 babel-jest         ^5.3.0  →       ^6.0.1 
 babelify           ^6.3.0  →       ^7.2.0 
 browserify        ^11.2.0  →      ^12.0.1 
 gulp-babel         ^5.2.1  →       ^6.1.1 
 gulp-modernizr      0.0.0  →  1.0.0-alpha 
 gulp-phpunit       ^0.9.1  →      ^0.10.1 
 jest-cli          ^0.5.10  →       ^0.8.1 
 react-icons        ^0.1.4  →       ^1.0.3 
 react-responsive   0.0.10  →        1.0.1 
 react-proxy        ^1.1.1  →       ^2.0.0 

Run with -u to upgrade package.json

The following dependencies are satisfied by their declared version range, but the installed versions are behind. You can install the latest versions without modifying your package file by using npm update. If you want to update the dependencies in your package file anyway, use --upgradeAll.

 es6-promise-polyfill   ^1.0.3  →   ^1.2.0 
 history               ^1.13.1  →  ^1.15.0 
 react-google-maps      ^4.3.1  →   ^4.7.0 
 react-router           ^1.0.0  →   ^1.0.2 
```

This sounds reasonable, as I ran 'npm update' a couple of days ago, and some packages has probably updated since then.

However, if I update npm-check-updates to the latest (2.5.4) again, I get this output:

```
 babel-core        ^5.8.25  →      ^6.3.15 
 babel-jest         ^5.3.0  →       ^6.0.1 
 babelify           ^6.3.0  →       ^7.2.0 
 browserify        ^11.2.0  →      ^12.0.1 
 gulp-babel         ^5.2.1  →       ^6.1.1 
 gulp-modernizr      0.0.0  →  1.0.0-alpha 
 gulp-phpunit       ^0.9.1  →      ^0.10.1 
 jest-cli          ^0.5.10  →       ^0.8.1 
 react-icons        ^0.1.4  →       ^1.0.3 
 react-responsive   0.0.10  →        1.0.1 
 react-proxy        ^1.1.1  →       ^2.0.0 

The following dependencies are satisfied by their declared version range, but the installed versions are behind. You can install the latest versions without modifying your package file by using npm update. If you want to update the dependencies in your package file anyway, use -a/--upgradeAll.

 autoprefixer                    ^6.0.3  →   ^6.1.2 
 browser-sync                   ^2.9.10  →  ^2.10.0 
 browserify-shim                ^3.8.10  →  ^3.8.11 
 counterpart                    ^0.17.0  →  ^0.17.1 
 csswring                        ^4.0.0  →   ^4.1.1 
 del                             ^2.0.2  →   ^2.2.0 
 es6-promise-polyfill            ^1.0.3  →   ^1.2.0 
 gulp-debug                      ^2.1.1  →   ^2.1.2 
 gulp-git                        ^1.5.0  →   ^1.6.0 
 gulp-sass                       ^2.0.4  →   ^2.1.0 
 gulp-uglify                     ^1.4.1  →   ^1.5.1 
 gulp-util                       ^3.0.6  →   ^3.0.7 
 history                        ^1.13.1  →  ^1.15.0 
 react                          ^0.14.2  →  ^0.14.3 
 react-addons-test-utils        ^0.14.2  →  ^0.14.3 
 react-addons-transition-group  ^0.14.2  →  ^0.14.3 
 react-addons-update            ^0.14.2  →  ^0.14.3 
 react-dom                      ^0.14.2  →  ^0.14.3 
 react-google-maps               ^4.3.1  →   ^4.7.0 
 react-router                    ^1.0.0  →   ^1.0.2 
 run-sequence                    ^1.1.4  →   ^1.1.5 
 sinon                          ^1.17.1  →  ^1.17.2 
 superagent                      ^1.4.0  →   ^1.5.0 
 watchify                        ^3.4.0  →   ^3.6.1 

Run with -u to upgrade package.json
```

For some reason, it thinks a lot more packages is behind, even though when I check the node_modules folder and the installed version for all of the packages, only the ones from the output of verson 2.5.1 is actually behind, the others are up to date. And this is also confirmed with the output out 'npm list --depth=0'

```
viking-nature-launch-2015@0.0.1 E:\WEB\viking-nature-launch-2015
├── autoprefixer@6.1.2
├── babel-core@5.8.34
├── babel-jest@5.3.0
├── babel-plugin-react-transform@1.1.1
├── babelify@6.4.0
├── bows@1.4.8
├── browser-sync@2.10.0
├── browserify@11.2.0
├── browserify-shim@3.8.11
├── counterpart@0.17.1
├── csswring@4.1.1
├── del@2.2.0
├── es6-promise-polyfill@1.1.1
├── es6-shim@0.33.13
├── geolib@2.0.18
├── google-map-react@0.9.3
├── gulp@3.9.0
├── gulp-babel@5.3.0
├── gulp-changed@1.3.0
├── gulp-composer@0.1.0
├── gulp-concat@2.6.0
├── gulp-debug@2.1.2
├── gulp-expect-file@0.0.7
├── gulp-filesize@0.0.6
├── gulp-filter@3.0.1
├── gulp-git@1.6.0
├── gulp-modernizr@0.0.0
├── gulp-notify@2.2.0
├── gulp-phpunit@0.9.3
├── gulp-plumber@1.0.1
├── gulp-postcss@6.0.1
├── gulp-rename@1.2.2
├── gulp-replace@0.5.4
├── gulp-rev@6.0.1
├── gulp-rev-all@0.8.22
├── gulp-rev-replace@0.4.2
├── gulp-sass@2.1.0
├── gulp-sequence@0.4.1
├── gulp-sourcemaps@1.6.0
├── gulp-task-listing@1.0.1
├── gulp-uglify@1.5.1
├── gulp-util@3.0.7
├── history@1.14.0
├── jest-cli@0.5.10
├── livereactload@2.1.0
├── localforage@1.3.0
├── lockdown@0.0.8-dev
├── lodash@3.10.1
├── lodash.clone@3.0.3
├── lodash.isequal@3.0.4
├── lodash.merge@3.3.2
├── lodash.throttle@3.0.4
├── object-assign@4.0.1
├── pretty-hrtime@1.0.1
├── pubsub-js@1.5.3
├── raf@3.1.0
├── react@0.14.3
├── react-addons-test-utils@0.14.3
├── react-addons-transition-group@0.14.3
├── react-addons-update@0.14.3
├── react-dom@0.14.3
├── react-google-maps@4.6.1
├── react-h5-video@1.0.0
├── react-icons@0.1.5
├── react-markdown@1.1.1
├── react-proxy@1.1.1
├── react-responsive@0.0.10
├── react-router@1.0.1
├── reflux@0.3.0
├── require-dir@0.3.0
├── run-sequence@1.1.5
├── shortid@2.2.4
├── sinon@1.17.2
├── superagent@1.5.0
├── superagent-bluebird-promise@2.1.1
├── vinyl-buffer@1.0.0
├── vinyl-source-stream@1.1.0
└── watchify@3.6.1

```

Does the issue make more sense now?

**Comment by apapirovski** [comment]
Having the same issue.

```
 babel         ^5.8.29  →       ^6.3.13 
 history        1.13.x  →        1.16.x 
 babel-core    ^5.8.33  →       ^6.3.17 
 babel-eslint   ^4.1.4  →  ^5.0.0-beta6 
 babel-loader   ^5.3.2  →        ^6.2.0 

The following dependencies are satisfied by their declared version range, but the installed versions are behind. You can install the latest versions without modifying your package file by using npm update. If you want to update the dependencies in your package file anyway, use -a/--upgradeAll.

 hapi                       ^11.0.2  →       ^11.1.2 
 react                      ^0.14.1  →       ^0.14.3 
 react-dom                  ^0.14.1  →       ^0.14.3 
 react-helmet                ^2.1.1  →        ^2.2.0 
 react-redux                 ^4.0.0  →        ^4.0.1 
 react-router                ^1.0.1  →        ^1.0.2 
 redux                       ^3.0.4  →        ^3.0.5 
 redux-router          ^1.0.0-beta3  →  ^1.0.0-beta5 
 redux-thunk                 ^1.0.0  →        ^1.0.2 
 clean-webpack-plugin        ^0.1.4  →        ^0.1.5 
 eslint-plugin-react        ^3.11.2  →       ^3.11.3 
```

```
├── autoprefixer@6.1.2
├── babel@5.8.34
├── babel-core@5.8.34
├── babel-eslint@4.1.6
├── babel-loader@5.4.0
├── clean-webpack-plugin@0.1.5
├── css-loader@0.23.0
├── css-mqpacker@4.0.0
├── dotenv@1.2.0
├── eslint@1.10.3
├── eslint-config-airbnb@2.0.0
├── eslint-plugin-react@3.11.3
├── estraverse@4.1.1
├── extract-text-webpack-plugin@0.9.1
├── file-loader@0.8.5
├── hammerjs@2.0.4
├── hapi@11.1.2
├── history@1.13.1
├── inert@3.2.0
├── isomorphic-fetch@2.2.0
├── lodash@3.10.1
├── node-geoip2@0.10.1
├── postcss@5.0.12
├── postcss-calc@5.0.0
├── postcss-color-function@2.0.0
├── postcss-conditionals@2.0.0
├── postcss-custom-media@5.0.0
├── postcss-custom-properties@5.0.0
├── postcss-custom-selectors@3.0.0
├── postcss-import@7.1.3
├── postcss-loader@0.8.0
├── postcss-mixins@3.0.2
├── postcss-nested@1.0.0
├── postcss-url@5.0.2
├── raf@3.1.0
├── raven@0.9.0
├── react@0.14.3
├── react-dom@0.14.3
├── react-helmet@2.2.0
├── react-redux@4.0.1
├── react-router@1.0.2
├── redux@3.0.5
├── redux-router@1.0.0-beta5
├── redux-thunk@1.0.2
├── schema-client@1.2.9
├── spdy@3.0.0
├── style-loader@0.13.0
├── url-loader@0.5.7
├── webpack@1.12.9
├── webpack-manifest-plugin@0.5.0
└── yar@5.0.0
```

**Comment by andersaloof** [comment]
I've debugged a bit and narrowed the issue down to the commit 3273e314fe294ebc74e0b3541eea61e1e8b77837.

The changes in this commit makes the .dependencies object in the resolve callback empty, and thus the module thinks no packages are installed and all packages are reported as behind.

Reverting the changes in this commit fixes the issue for my part and the list of satisfied packages that are behind is reduced from 26 to 7, which is correct since I've not run npm update since I reported this issue.

**Comment by strawbrary** [comment]
I've hit the same issue, as @andersaloof uncovered it looks like a regression since v2.5.3. Downgrading to v2.5.2 fixes the issue.

**Comment by metaraine** [comment]
@andersaloof Thank you! It's so great to have people who are willing to look at the source code. It is tough to be the only contributor. Thanks for your time.

It was a big regression. I reverted it and added a unit test to prevent this from happening in the future.

Published to v2.5.5.
